controllers/storagecluster/odfinfoconfig.go

```diff
@@ -0,0 +1,294 @@
+	"strings"
+	"sync"
+
+	operatorsv1alpha1 "github.com/operator-framework/api/pkg/operators/v1alpha1"
+	ocsv1 "github.com/red-hat-storage/ocs-operator/api/v4/v1"
+	"github.com/red-hat-storage/ocs-operator/v4/controllers/util"
+	rookCephv1 "github.com/rook/rook/pkg/apis/ceph.rook.io/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/klog/v2"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/yaml"
+)
+
+type connectedClient struct {
+	NamespacedName types.NamespacedName `yaml:"metadata"`
+	ClusterID      string               `yaml:"clusterId"`
+}
+type infoStorageCluster struct {
+	NamespacedName          types.NamespacedName `yaml:"metadata"`
+	StorageProviderEndpoint string               `yaml:"storageProviderEndpoint"`
+	CephClusterFSID         string               `yaml:"cephClusterFSID"`
+}
+
+type OdfInfoData struct {
+	Version           string             `yaml:"version"`
+	DeploymentType    string             `yaml:"deploymentType"`
+	Clients           []connectedClient  `yaml:"clients"`
+	StorageCluster    infoStorageCluster `yaml:"storageCluster"`
+	StorageSystemName string             `yaml:"storageSystemName"`
+	IsDROptimized     bool               `yaml:"isDROptimized"`
+}
+
+const (
+	odfInfoKeySuffix          = "config.yaml"
+	odfDeploymentTypeExternal = "external"
+	odfDeploymentTypeInternal = "internal"
+	rookCephMonSecretName     = "rook-ceph-mon"
+	fsidKey                   = "fsid"
+	odfOperatorNamePrefix     = "odf-operator"
+	odfInfoConfigMapName      = "odf-info"
+	odfInfoMapKind            = "ConfigMap"
+)
+
+type odfInfoConfig struct{}
+
+var mutex sync.RWMutex
+
+// ensureCreated ensures that a ConfigMap resource exists with its Spec in
+// the desired state.
+func (obj *odfInfoConfig) ensureCreated(r *StorageClusterReconciler, storageCluster *ocsv1.StorageCluster) (reconcile.Result, error) {
+	operatorNamespace, err := util.GetOperatorNamespace()
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	odfInfoConfigMap := &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      odfInfoConfigMapName,
+			Namespace: operatorNamespace,
+		},
+	}
+
+	mutex.Lock()
+	defer mutex.Unlock()
+	_, err = ctrl.CreateOrUpdate(r.ctx, r.Client, odfInfoConfigMap, func() error {
+		if err = ctrl.SetControllerReference(storageCluster, odfInfoConfigMap, r.Scheme); err != nil {
+			return err
+		}
+		r.Log.Info("Creating or updating odf-info config map", odfInfoMapKind, klog.KRef(storageCluster.Namespace, odfInfoConfigMapName))
+		odfInfoKey := obj.getOdfInfoKeyName(storageCluster)
+		var data map[string]string
+		if err = yaml.Unmarshal([]byte(odfInfoConfigMap.Data[odfInfoKey]), data); err != nil {
+			return fmt.Errorf("failed to get odf-info config map data: %v", err)
+		}
```

Comment:
This can explode as there might not even be a key here (in case this is the first reconcile to build that key). 
But regardless, nothing in this function needs the old data. The function rebuilds a new version of the data an overwrites the key. So in that case, why do we need to even unmarshal existing key content? 

```suggestion
```